src/logger.rs

```diff
@@ -19,14 +44,24 @@ impl Log for Logger {
             return;
         }
 
-        let formatted = format!("{} - {} - {}", Local::now().format("%Y-%m-%d %H:%M:%S"), record.level(), record.args());
+        let timestamp = Local::now().format("%Y-%m-%d %H:%M:%S");
+
+        eprintln!("{}", record.args());
 
-        eprintln!("{}", formatted);
+        let message = LogMessage {
+            timestamp: timestamp.to_string(),
+            level: record.level(),
+            context: self.context.clone(),
+            message: format!("{}", record.args())
+        };
 
         self.tcp_stream.as_ref().map(|ref stream| {
             stream.lock()
                 .map_err(|e| e.to_string())
-                .and_then(|mut stream| writeln!(*stream, "{}", formatted).map_err(|e| e.to_string()))
+                .and_then(|mut stream| {
+                    serde_json::to_writer_pretty(&(*stream), &message).map_err(|e| e.to_string())?;
```

Comment:
Isn't & then * redundant?

Reply:
Oddly enough, no it isn't. The error message that you get if you take it out is:

```
serde_json::to_writer_pretty(stream, &message).map_err(|e| e.to_string())?;
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::io::Write` is not implemented for `std::sync::MutexGuard<'_, std::net::TcpStream>`
```

Dereferencing it gets you the actual stream, not the MutexGuard. Then the function wants to borrow the actual stream, so you pass a reference to it.